core/src/main/java/org/pentaho/di/core/row/RowMeta.java

```diff
@@ -0,0 +1,1333 @@
+    while ( searchValueMeta( name ) != null ) {
+      index++;
+      name = valueMeta.getName() + "_" + index;
+    }
+
+    // Create a copy of the valueMeta object to make sure we don't rename any other value meta objects.
+    // It's only being renamed because of the addition to THIS row metadata object, not another.
+    //
+    ValueMetaInterface copy = valueMeta.clone();
+
+    // OK, this is the new name and origin to pick
+    //
+    copy.setName( name );
+    if ( originStep != null ) {
+      copy.setOrigin( originStep );
+    }
+    return copy;
+  }
+
+  /**
+   * Get an array of the names of all the Values in the Row.
+   *
+   * @return an array of Strings: the names of all the Values in the Row.
+   */
+  @Override
+  public String[] getFieldNames() {
+    lock.readLock().lock();
+    try {
+      String[] retval = new String[ size() ];
+
+      for ( int i = 0; i < size(); i++ ) {
+        String valueName = getValueMeta( i ).getName();
+        retval[i] = valueName == null ? "" : valueName;
+      }
+
+      return retval;
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Write ONLY the specified data to the outputStream
+   *
+   * @throws KettleFileException in case things go awry
+   */
+  @Override
+  public void writeData( DataOutputStream outputStream, Object[] data ) throws KettleFileException {
+    lock.readLock().lock();
+    try {
+      // Write all values in the row
+      for ( int i = 0; i < size(); i++ ) {
+        getValueMeta( i ).writeData( outputStream, data[ i ] );
+      }
+
+      // If there are 0 values in the row, we write a marker flag to be able to detect an EOF on the other end (sockets
+      // etc)
+      //
+      if ( size() == 0 ) {
+        try {
+          outputStream.writeBoolean( true );
+        } catch ( IOException e ) {
+          throw new KettleFileException( "Error writing marker flag", e );
+        }
+      }
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Write ONLY the specified metadata to the outputStream
+   *
+   * @throws KettleFileException in case things go awry
+   */
+  @Override
+  public void writeMeta( DataOutputStream outputStream ) throws KettleFileException {
+    lock.readLock().lock();
+    try {
+      // First handle the number of fields in a row
+      try {
+        outputStream.writeInt( size() );
+      } catch ( IOException e ) {
+        throw new KettleFileException( "Unable to write nr of metadata values", e );
+      }
+
+      // Write all values in the row
+      for ( int i = 0; i < size(); i++ ) {
+        getValueMeta( i ).writeMeta( outputStream );
+      }
+    } finally {
+      lock.readLock().unlock();
+    }
+
+  }
+
+  public RowMeta( DataInputStream inputStream ) throws KettleFileException, SocketTimeoutException {
+    this();
+
+    int nr;
+    try {
+      nr = inputStream.readInt();
+    } catch ( SocketTimeoutException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)